### ISSUE TITLE
missing capability checks

### ISSUE THREAD

**Issue by danmarsden** [DevEticeo/moodle-mod_glossaryfocus#4]
https://github.com/academic-moodle-cooperation/moodle-mod_consentform/blob/master/styles.css#L39

your external functions must check if the current logged in user can view the glossary being passed.

this is a blocker for approval in the plugins db.

**Comment by DevEticeo** [comment]
Hi Dan, 
Many thanks for your feedbacks.
Regarding this specific point. Actualy, the aim of this plugin is to give access to a given glossary article without the need to have access to the glossary from where the article is picked up.
The vision is to have one unique glossary somewhere on the lms and to show several article from this one into other courses.
In this way, all articles are stored in the glossary and we do not have to enroll each learners on the course hosting this glossary. Article contents are directly extracted from the data base.
I hope I'm clear enough to allow this blocking point to be unlocked.
regards

**Comment by danmarsden** [comment]
Think about the situation where a teacher creates a glossary inside their course but isn't planning to use it with glossaryfocus - it might have sensitive information or should only be available to users inside the course/

your external lib functions allow all users on the site to bypass this by calling the external function you've written "get_words" passing in any glossary id they want which could be contained on any course in the site, including courses they should not be able to access.

I note that I pasted the wrong link above - here's the right one which shows it querying the core glossary tables (not your custom glossaryfocus plugin)
https://github.com/DevEticeo/moodle-mod_glossaryfocus/blob/main/classes/external.php#L80-L83

I'm not completely clear on why you've implemented it like this or why you need it to work this way, but it bypasses the security protections implemented within the mod_glossary plugin, allowing all users on the site to access information from all glossary instances on the site - no matter if they are hidden from all users, not enrolled etc.

So... no... I don't think we'd allow approval in this state - you might need to rewrite how this works a bit.

**Comment by DevEticeo** [comment]
Many thanks for your quick response.
I see the security point. We will work on it and propose something suiting the restriction.

**Comment by DevEticeo** [comment]
what about if we restrict only non hidden global glossaries to be used throught the plugin ?

**Comment by danmarsden** [comment]
Yeah, ideally something that checks, can this user see this glossary would be ideal. Like in glossary view here https://github.com/moodle/moodle/blob/4ce642e8ba568ca50dd71f77c0e231a5676ec8c1/mod/glossary/view.php#L53

**Comment by DevEticeo** [comment]
Thanks a lot, we will add the capability check.
Regards, 
Lilian

**Comment by danmarsden** [comment]
Hi Lilian, 

I can't see  a capability check in the "get_words" external function yet?
https://github.com/DevEticeo/moodle-mod_glossaryfocus/blob/main/classes/external.php#L69

that's the one that still needs a check - "can this user, view the glossary that it's asking for information about."

**Comment by DevEticeo** [comment]
Hi Dan,
In fact we put the capablily check when we display the words to the students (in the function render_glossaryfocus_words https://github.com/DevEticeo/moodle-mod_glossaryfocus/blob/06b310fd0dd8abe05336247e5c6b8e61158835b9/renderer.php#L35).

The mod_glossaryfocus\external::get_words function is used only on the backoffice side when the teacher creates the focus glossary (it is this function that displays the words that match the letters entered in the input) and even if we don't check the capability, we only return the words in a global glossary or in the current course.

But we can also add it in this function, like we say in French, "it doesn't eat bread". :)

Gabrielle Guevara (Lilian is on vacation)

**Comment by danmarsden** [comment]
Yeah, but all that does is hide the link. Just because you use this webservice behind the scenes... doesn't mean that your users will. Don't trust the users right? They might see this webservice and hit it directly rather than via the links in your page. :-)

**Comment by DevEticeo** [comment]
Yes good remarks indeed! :) I correct that right away. Thanks!
